### PR TITLE
Fix the issue "Endless exception dialogues with illegal char in file path"

### LINKED ISSUE #6859: Invisible buttons for File field
JabRef 5.1--2020-08-30--e023aa0
Windows 10 10.0 amd64 
Java 14.0.2

- [x ] **Mandatory**: I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

The buttons near the File field are not visible (screenshot 1). If the File entry has problems all the following fields disappear (screenshot 2).

1. Open a library containing the following entry:
```
@InCollection{Favre2004,
  Title                    = {{AVS and AVS/Express}},
  Author                   = {Jean Favre and Mario Valle},
  Booktitle                = {{The Visualization Handbook}},
  Publisher                = {Academic Press},
  Year                     = {2004},
  Editor                   = {Chuck Hansen and Chris Johnson},
  Month                    = dec,
  Pages                    = {655--672},
  File                     = {Paper:file\:///D\:\\Books\\Publications\\ch33%20favre.pdf:PDF},
  Keywords                 = {books},
  Mvpubtype                = {books},
  Langid                   = {english},
  Url                      = {http://www.elsevierdirect.com/product.jsp?isbn=9780123875822}
}
```
2. Select the entry and an exception appears (exception stack below):
```
Illegal char <:> at index 4: file:///D:/Books/Publications/ch33%20favre.pdf
```
Note this entry pertains to a library created in a very old version of JabRef.
3. Go to the General tab and notice the buttons are not visible.
4. Click on the File field and all the following fields disappear.
5. Now try to reload the entry selecting the file with the leftmost invisible button.
6. Nothing appears in the field

Missing buttons:
image

After clicking on the File field:
image

Exception text:
```
java.nio.file.InvalidPathException: Illegal char <:> at index 4: file:///D:/Books/Publications/ch33%20favre.pdf
	at java.base/sun.nio.fs.WindowsPathParser.normalize(Unknown Source)
	at java.base/sun.nio.fs.WindowsPathParser.parse(Unknown Source)
	at java.base/sun.nio.fs.WindowsPathParser.parse(Unknown Source)
	at java.base/sun.nio.fs.WindowsPath.parse(Unknown Source)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(Unknown Source)
	at java.base/java.nio.file.Path.of(Unknown Source)
	at org.jabref@5.1.60000/org.jabref.gui.fieldeditors.LinkedFileViewModel.isGeneratedNameSameAsOriginal(Unknown Source)
	at org.jabref@5.1.60000/org.jabref.gui.fieldeditors.LinkedFilesEditor.createContextMenuForFile(Unknown Source)
	at org.jabref@5.1.60000/org.jabref.gui.util.ViewModelListCellFactory$1.updateItem(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.ListCell.updateItem(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.ListCell.indexChanged(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.IndexedCell.updateIndex(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.skin.VirtualFlow.setCellIndex(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.skin.VirtualFlow.getCell(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.skin.VirtualFlow.getCellLength(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.skin.VirtualFlow.computeViewportOffset(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.control.skin.VirtualFlow.layoutChildren(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Parent.layout(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Scene.doLayoutPass(Unknown Source)
	at org.jabref.merged.module@5.1.60000/javafx.scene.Scene$ScenePulseListener.pulse(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.Toolkit.runPulse(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.Toolkit.firePulse(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at org.jabref.merged.module@5.1.60000/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at org.jabref.merged.module@5.1.60000/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

### PR DESCRIPTION
**Pull request by Fu188**
If the user open a library including an illegal file path, the application will crash in the original version (endless exception dialogues). In this case, I add an exception handling to deal with this problem and prompt user "the file path is invalid"

Fixes #6859

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

**Testcase**

@InCollection{Favre2004,
Title = {{AVS and AVS/Express}},
Author = {Jean Favre and Mario Valle},
Booktitle = {{The Visualization Handbook}},
Publisher = {Academic Press},
Year = {2004},
Editor = {Chuck Hansen and Chris Johnson},
Month = dec,
Pages = {655--672},
File = {Paper:file:///D:\Books\Publications\ch33%20favre.pdf:PDF},
Keywords = {books},
Mvpubtype = {books},
Langid = {english},
Url = {http://www.elsevierdirect.com/product.jsp?isbn=9780123875822}
}

### CODE REVIEW

File: src/main/java/org/jabref/logic/importer/util/FileFieldParser.java [hunk 1]
```diff
@@ -77,17 +82,25 @@ private static LinkedFile convert(List<String> entry) {
...
-        } else if (!field.getDescription().isEmpty() && field.getLink().isEmpty() && field.getFileType().isEmpty()) {
-            field = new LinkedFile("", Path.of(field.getDescription()), "");
+            // link is only mandatory field
+            if (field.getDescription().isEmpty() && field.getLink().isEmpty() && !field.getFileType().isEmpty()) {
+                field = new LinkedFile("", Path.of(field.getFileType()), "");
+            } else if (!field.getDescription().isEmpty() && field.getLink().isEmpty() && field.getFileType().isEmpty()) {
+                field = new LinkedFile("", Path.of(field.getDescription()), "");
+            }
+            entry.clear();
+        }catch (InvalidPathException exception){
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/util/FileFieldParser.java:97]: Checkstyle: Space after curly bracket (`} catch (...`

File: src/main/java/org/jabref/logic/importer/util/FileFieldParser.java [hunk 2]
```diff
@@ -77,17 +82,25 @@ private static LinkedFile convert(List<String> entry) {
...
+            DefaultTaskExecutor.runInJavaFXThread(() ->
+                    JabRefGUI.getMainFrame()
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/util/FileFieldParser.java:99]: I believe this causes a architecture violation. Can you instead catch the InvalidPathException in the closest gui package caller method?

File: src/main/java/org/jabref/logic/importer/util/FileFieldParser.java [hunk 3]
```diff
@@ -56,7 +59,9 @@
         }
 
         if (!entry.isEmpty()) {
-            files.add(convert(entry));
+            if(convert(entry)!=null){
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/util/FileFieldParser.java:62]: Checkstyle: Keep your spaces

**Review by calixtus — CHANGES_REQUESTED** [review]
Hi @Fu188, thank you for your interest in JabRef development.
I noticed you have some checkstyle issues. Please make sure you have configured your workspace properly according to our guidelines (https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace).
There is also an issue with your changes: We try to distinguish in JabRef a model, a logic and a gui package. Although it already works out most of the time, we still have some work to do. Unfortunately the idea to call to the DialogService in a logic package violates this architecture. We would be happy if you'd like to rework your code. My idea would be to catch the exceptions everytime a call to `FileFieldParser::parse` is made.

**Comment by calixtus** [comment]
This seems to be a duplication of #7386. I'll close the other.

File: src/main/java/org/jabref/logic/importer/util/FileFieldParser.java [hunk 4]
```diff
@@ -56,7 +59,9 @@
...
+                files.add(convert(entry));
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/util/FileFieldParser.java:63]: Maybe introduce a new variable instead calling `convert(entry)` twice?

File: src/main/java/org/jabref/logic/importer/util/FileFieldParser.java (hunk 2 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/importer/util/FileFieldParser.java:99]: Shouldn't this just return a "unparsed" LinkedFile? @koppor mabye can clarify that point.

**Comment by Fu188** [comment]
I have pull request the revised version. Since the original version will not have a dialog to prompt user if the file path has no illegal char but invalid, I keep the same way if the file path has illegal char. I just catch the exception and ignore it to prevent the application crashing. By the way, I found it is hard to catch the exceptions everytime a call to FileFieldParser::parse is made due to it is implemented in `map` function. Above all, I think we only need to catch the exception. I hope you can accept my pull request, it is very important for me!

File: src/main/java/org/jabref/logic/importer/util/FileFieldParser.java [hunk 5]
```diff
@@ -2,10 +2,13 @@
 
 import java.net.MalformedURLException;
 import java.net.URL;
+import java.nio.file.InvalidPathException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 
+import org.jabref.gui.JabRefGUI;
+import org.jabref.gui.util.DefaultTaskExecutor;
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/util/FileFieldParser.java:11]: Two unused imports

File: CHANGELOG.md [hunk 6]
```diff
@@ -35,6 +35,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where the Harvard RTF exporter did not use the new authors formatter and therefore did not export "organization" authors correctly. [4508](https://github.com/JabRef/jabref/issues/4508)
 - We fixed an issue where the field `urldate` was not exported to the corresponding fields `YearAccessed`, `MonthAccessed`, `DayAccessed` in MS Office XML [#7354](https://github.com/JabRef/jabref/issues/7354)
 - We fixed an issue where the password for a shared SQL database was only remembered if it was the same as the username [#6869](https://github.com/JabRef/jabref/issues/6869)
+- We fixed an issue where endless exception dialogues with illegal char in file path. [#7381](https://github.com/JabRef/jabref/issues/7381)
```
Comment (calixtus) [inline CHANGELOG.md:38]:
```suggestion
- We fixed an issue with endless exception dialogues appearing if the file path includes an illegal char. [#7381](https://github.com/JabRef/jabref/issues/7381)
```

**Review by calixtus — CHANGES_REQUESTED** [review]
Two little remarks. Thanks for the quick follow up.

**Comment by Fu188** [comment]
> Two little remarks. Thanks for the quick follow up.

Thanks for your suggestion. I have changed them and pulled the new version. Please check it.

**Pull request closed without merging**